### PR TITLE
KOGITO-8229: Dashbuilder ECharts Native

### PR DESCRIPTION
**Pull request by jesuino**
Bringing ECharts as the default chart library for the following types:

BAR, LINE, AREA, BUBBLE, PIE and METRIC.

C3 will remains as an alternative

The main reason for this change is that ECharts has a better native support for dark mode and easier to customize

For this reason the following attributes are now supported:

**general.mode**: dark or light;
**extraConfiguration**: Additional JSON configuration based on ECharts [Option](https://echarts.apache.org/en/option.html#title) object. This configuration overrides the configuration used by Dashbuilder and allow users to build more complex visualization. The only limitation with this feature is that the extra configuration is JSON, not a Javascript object, so using callbacks on formatter does not work.

https://user-images.githubusercontent.com/359820/201425099-9f5f365f-289f-4c20-a74d-8b785676dc1a.mp4

A new chart type is also now supported. SCATTERCHART is a new type and it was "for free" on echarts, so we now support it.

A Basic echarts usage is as following. Notice it is in dark mode and contains a title

```
datasets:
- uuid: test
  content: >-
            [
                ["A", 1, 3],
                ["B", 2, 5],
                ["C", 3, 1]
            ]
pages:
- components:
    - settings:
        type: BARCHART
        general:
            title: Simple Bar Chart
            mode: dark
        lookup:
          uuid: test
```

image

You can change to c3, but dark mode will be ignored. Just add a field called "renderer" with the value "c3"

```
datasets:
- uuid: test
  content: >-
            [
                ["A", 1, 3],
                ["B", 2, 5],
                ["C", 3, 1]
            ]
pages:
- components:
    - settings:
        type: BARCHART
        renderer: c3
        general:
            title: Simple Bar Chart
            mode: dark
        lookup:
          uuid: test
```

Which renders to the following:

image

Finally, some advantages of using echarts we added new properties:

* **export.png**: adds an icon to download the PNG of a chart
* **general.allowEdit**: an icon when clicked open the data in text mode and allow users to edit the data used by the chart
* **bubble.minSize**: a min size for bubble chart
* **bubble.maxSize**: a min size for bubble chart
* * **bubble.color**: Select the bubble color

To run this build until "dashbuilder-editor" and run the project as a web application. This is already live on my test editor: https://jesuino.github.io/dashbuilder-client/

**************************

https://issues.redhat.com/browse/KOGITO-8252

Since the very reason for this PR is dark mode, so this PR also includes dark mode. To achieve this new feature we added support to a new top level attribute called "global" and it supports two fields:

* **global.mode**: Can be `dark` or `light` - it is the mode for the entire report;
* **global.settings**: It is a global settings object. It has a global setting that can be applied to all components on the dashboard, except for the `lookup`, the lookup is not a global configuration. The  component declared in the layout can override the global setting

### CODE REVIEW

File: packages/dashbuilder-component-table/src/FilteredTableComponent.tsx [hunk 1]
```diff
@@ -36,10 +38,15 @@ export function FilteredTableComponent(props: Props) {
       const selectableParam = params.get("selectable") === "true";
       const filterColumnParam = +params.get("filterColumn") || 0;
       const columnStr = params.get("alertColumn");
+      const hideHeader = params.get("hideHeader") === "true";
+
+      // TODO: Use a proper Props object..
```
Comment (caponetto) [inline packages/dashbuilder-component-table/src/FilteredTableComponent.tsx]: Leftover?

File: packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/pom.xml [hunk 2]
```diff
@@ -0,0 +1,185 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (caponetto) [inline packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/pom.xml:1]: Missing copyright header.

File: packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/java/org/dashbuilder/renderer/echarts/client/EChartsAbstractDisplayer.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    @Override
+    protected void updateVisualization() {
+        if (dataSet.getRowCount() == 0) {
+            getView().noData();
+        } else {
+            updateVisualizationWithData();
+        }
+    }
+
+    // TODO: need some action to clear filter
```
Comment (caponetto) [inline packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/java/org/dashbuilder/renderer/echarts/client/EChartsAbstractDisplayer.java]: Leftover?

File: packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/resources/org/dashbuilder/renderer/EChartsRenderer.gwt.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (caponetto) [inline packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/resources/org/dashbuilder/renderer/EChartsRenderer.gwt.xml:1]: Missing copyright header.

**Review by caponetto — COMMENTED** [review]
Very nice, @jesuino! Left some minor comments inline.

**Comment by jesuino** [comment]
Hello @caponetto 

Thanks for your review, updated the code with your suggestions

**Comment by caponetto** [comment]
Hey @jesuino, the code looks good to me but there are some failing integrations tests.
Can you please check? I think the test is trying to get C3 stuff where it doesn't exist anymore.

File: packages/dashbuilder-component-table/static/index.css [hunk 5]
```diff
@@ -0,0 +1,44 @@
+.selected-row {
```
Comment (caponetto) [inline packages/dashbuilder-component-table/static/index.css:16]: Missing copyright header.

File: packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/java/org/dashbuilder/renderer/echarts/client/ChartBootstrapParams.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ */
+package org.dashbuilder.renderer.echarts.client;
+
+import java.util.Objects;
+
+import org.dashbuilder.displayer.Mode;
+import org.dashbuilder.renderer.echarts.client.js.ECharts.Renderer;
+
+/**
+ * @author wsiqueir
```
Comment (caponetto) [inline packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/java/org/dashbuilder/renderer/echarts/client/ChartBootstrapParams.java]: AFAIK, we don't use `@author` anymore.

**Review by caponetto — APPROVED** [review]
Thanks, @jesuino! LGTM

**Comment by barboras7** [comment]
Hi @jesuino,
is this the final version with all the commits? I did some basic manual check of the PR last week and it did look good, but I want to take a final look at the UI with all the commits before it gets merged.

**Comment by jesuino** [comment]
Hello @barboras7 

It is the final version now - I added dark mode on top of it because we would double our work to review another PR

If you see new commits it is just because I am testing the dark mode and sometimes I find small bugs

Thanks!

**Comment by jesuino** [comment]
Just for information

cypress is failing with a weird error due this: https://github.com/cypress-io/cypress/issues/24652

I updated cypress to 1.10.0 and it worked

**Review by caponetto — APPROVED** [review]

File: packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/java/org/dashbuilder/renderer/echarts/client/EChartsRenderer.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                return beanManager.lookupBean(EChartsXYChartDisplayer.class).newInstance();
+            case BUBBLECHART:
+                return beanManager.lookupBean(EChartsBubbleChartDisplayer.class).newInstance();
+            case PIECHART:
+                return beanManager.lookupBean(EChartsPieChartDisplayer.class).newInstance();
+            case METERCHART:
+                return beanManager.lookupBean(EChartsMeterChartDisplayer.class).newInstance();
+            case MAP:
+            default:
+                throw new IllegalArgumentException("Type not supported by ECharts");
```
Comment (barboras7) [inline packages/dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-echarts/src/main/java/org/dashbuilder/renderer/echarts/client/EChartsRenderer.java]: I'd maybe extend this to ".... by Echarts, use C3 renderer/charts instead"

**Review by barboras7 — COMMENTED** [review]
Hi @jesuino!

Leaving just some minor comments. 

- IMHO, it would be nice to also fix the description in the kitchensink example since C3 are no longer default charts, see [kitchensink line here](https://github.com/kiegroup/kie-tools/blob/0662ba38408c23017eab69ad6e459ea54ab6bb46/examples/dashbuilder/kitchensink/Kitchensink.dash.yml#L537)

- I like the dark mode, however if you google something like "charts dark mode", you don't only get specific chart in the dark mode but everything rendered is in dark. I understand that this would require much more work, but maybe something to think about for future. It looks a bit strange to just to have one dark chart with everything white around it.

**Comment by jesuino** [comment]
Hello @barboras7 

Thank you so much for your review.

>> I like the dark mode, however if you google something like "charts dark mode", you don't only get specific chart in the dark mode but everything rendered is in dark. I understand that this would require much more work, but maybe something to think about for future. It looks a bit strange to just to have one dark chart with everything white around it.

The local declaration of mode is override by the global mode., see this example:

```
global:
  mode: dark
datasets:
  - uuid: test
    content: >-
                [
                    ["A", 1],
                    ["B", 2],
                ]
pages:
  - components:
      - settings:
          type: BARCHART
          lookup:
            uuid: test
```
This renders to the following

image


You can see a full example here:

https://jesuino.github.io/dashbuilder-client/?import=https://raw.githubusercontent.com/jesuino/dashbuilder-yaml-samples/main/random/fifa_2022_goals.dash.yaml#editor=false&preview=true

**Review by barboras7 — APPROVED** [review]
Thanks for the fixes @jesuino, looks great! :tada:

**Pull request merged**